extensions/haskell_toolchains.bzl

```diff
@@ -0,0 +1,165 @@
+)
+
+def _all_toolchains_impl(rctx):
+    content = "\n".join(rctx.attr.toolchains)
+    rctx.file("BUILD.bazel", content = content)
+
+_all_toolchains = repository_rule(
+    implementation = _all_toolchains_impl,
+    attrs = {
+        "toolchains": attr.string_list(
+            doc = "All the `toolchain(...)` declarations for the bindist toolchains as string",
+        ),
+    },
+    doc = """ Used to generate the `all_bindist_toolchains` external repository.
+    We can then invoque `register_toolchains("@all_bindist_toolchains//:all")` in the MODULE.bazel file.
+    """,
+)
+
+def _haskell_toolchains_impl(mctx):
+    # We gather the declarations of all the toolchains in the `toolchain_declarations` list
+    # in order to write them to the `@all_bindist_toolchains` repository.
+
+    # The code could be simplified a bit by registering aliases to the
+    # `toolchain` rules, once the following issue is resolved
+    # https://github.com/bazelbuild/bazel/issues/16298
+    toolchain_declarations = []
+
+    toolchain_names = []
+    found_bindists = False
+    for module in mctx.modules:
+        for bindist_tag in module.tags.bindist:
+            name = "bindist_{}_{}_{}".format(module.name, module.version, bindist_tag.name)
+            if name in toolchain_names:
+                fail(
+                    """Module "{module}~{version}" used the "bindist" tag twice with the "{tag_name}" name.""".format(
+                        tag_name = bindist_tag.name,
+                        module = module.name,
+                        version = module.version,
+                    ),
+                )
+            else:
+                toolchain_names.append(name)
+            ghc_bindist(
+                name = name,
+                version = bindist_tag.version,
+                target = bindist_tag.target,
+                ghcopts = bindist_tag.ghcopts,
+                haddock_flags = bindist_tag.haddock_flags,
+                repl_ghci_args = bindist_tag.repl_ghci_args,
+                cabalopts = bindist_tag.cabalopts,
+                locale = bindist_tag.locale,
+                register = False,
+            )
+            toolchain_declarations.append(
+                ghc_bindist_toolchain_declaration(
+                    target = bindist_tag.target,
+                    bindist_name = name,
+                    toolchain_name = name,
+                ),
+            )
+
+        if len(module.tags.bindists) > 1:
+            fail(
+                """Module "{module}~{version}" used the "bindists" tag more than once.""".format(
+                    module = module.name,
+                    version = module.version,
+                ),
+            )
+        for bindists_tag in module.tags.bindists:
+            # We only consider the first `bindists` tag, because subsequent
+            # ones would have the same constraints and lower priority.
+            if not found_bindists:
```

Comment:
Could this be simplified to something along the lines of the following?
```
if module.tags.bindists:
    bindists_tag = module.tags.bindists[0]
    ...
```
I.e. drop the `found_bindists` boolean.